### PR TITLE
Add strategy patterns to connect-rtc-js

### PR DESCRIPTION
**Pull request by siwei-w**
- Import @citrix/ucsdk to achieve the audio optimization for Citrix cloud desktops
- Introduce Strategy patterns for using different WebRTC SDK
- Add unit test for Strategy patterns

### Test
- Tested in Citrix and non-Citrix environment with embedded, custom, default CCP

### CODE REVIEW

File: src/js/strategies/CitrixVDIStrategy.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        rtcSession._pc.setRemoteDescription(answerSessionDescription, () => {
+            console.log(`ConnectRTC - rtc_session - AcceptState - rtcSession._pc.setRemoteDescription - success`);
+            var remoteCandidatePromises = Promise.all(self._candidates.map(function (candidate) {
+                var remoteCandidate = self._createRemoteCandidate(candidate);
+                self.logger.info('Adding remote candidate', remoteCandidate);
+                console.log(`ConnectRTC - rtc_session - AcceptState - rtcSession._pc.setRemoteDescription - Adding remote candidate - remoteCandidate`, remoteCandidate);
+                return rtcSession._pc.addIceCandidate(remoteCandidate);
+            }));
+            remoteCandidatePromises.catch(reason => {
+                console.log(`ConnectRTC - rtc_session - AcceptState - rtcSession._pc.setRemoteDescription - Adding remote candidate - failed`, reason);
```
Comment (yiming-amzn) [inline src/js/strategies/CitrixVDIStrategy.js:63]: we usually only get CCP log for investigation, console log are not usually shared. but is the reason not using logger here is due to some limitations?

File: src/js/strategies/CitrixVDIStrategy.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
+import CCPInitiationStrategyInterface from "./CCPInitiationStrategyInterface";
+import {hitch, isCitrixWebRTCSupported} from "../utils";
+import {FailedState} from "../rtc_session";
+import {RTC_ERRORS} from "../rtc_const";
+import "@citrix/ucsdk/CitrixWebRTC";
+
+export default class CitrixVDIStrategy extends CCPInitiationStrategyInterface {
```
Comment (yiming-amzn) [inline src/js/strategies/CitrixVDIStrategy.js:7]: are you planning to add unit test for these 2 strategy classes?

Comment (siwei-w) [reply src/js/strategies/CitrixVDIStrategy.js:7]: Yeah, will add more unit test causes to cover these.

File: src/js/strategies/CitrixVDIStrategy.js (hunk 1 above)
Comment (siwei-w) [reply src/js/strategies/CitrixVDIStrategy.js:63]: As the CCP logger is not initialized  in this class, we can either initialize CCP logger in this class or remove these console logs.

File: src/js/rtc_const.js [hunk 3]
```diff
@@ -34,13 +34,19 @@ export const BYE_METHOD_NAME = "bye";
 export const RTC_PEER_CONNECTION_CONFIG = {
     iceTransportPolicy: 'relay',
     rtcpMuxPolicy: 'require',
-    bundlePolicy: 'balanced'
+    bundlePolicy: 'balanced',
```
Comment (Vignendra) [inline src/js/rtc_const.js:37]: How do these configs affect non-citrix calls? Why are enforcing this on all calls?

File: src/js/rtc_peer_connection_factory.js [hunk 4]
```diff
@@ -1,21 +1,26 @@
...
 import {
-    CHROME_SUPPORTED_VERSION,
     DEFAULT_ICE_CANDIDATE_POOL_SIZE,
     NETWORK_CONNECTIVITY_CHECK_INTERVAL_MS,
     RTC_PEER_CONNECTION_CONFIG,
     RTC_PEER_CONNECTION_IDLE_TIMEOUT_MS,
     RTC_PEER_CONNECTION_OPTIONAL_CONFIG,
     SOFTPHONE_ROUTE_KEY
 } from './rtc_const'
+import CCPInitiationStrategyInterface from "./strategies/CCPInitiationStrategyInterface";
```
Comment (Vignendra) [inline src/js/rtc_peer_connection_factory.js:10]: CCPInitiationStrategyInterface - Is the naming correct? Would SoftphoneManagerInitiationStrategyInterface make more sense?

File: src/js/rtc_session.js [hunk 5]
```diff
@@ -119,14 +121,14 @@ export class GrabLocalMediaState extends RTCSessionState {
...
-        return navigator.mediaDevices.getUserMedia(constraints);
+        return this._rtcSession._strategy._gUM(constraints);
     }
 }
 export class CreateOfferState extends RTCSessionState {
     onEnter() {
         var self = this;
         var stream = self._rtcSession._localStream;
-        self._rtcSession._pc.addStream(stream);
+        self._rtcSession._strategy.createOfferStateOnEnter(self._rtcSession._pc, stream);
```
Comment (Vignendra) [inline src/js/rtc_session.js:131]: This function naming is very confusing - we are already inside OnEnter method and are calling onEnter again on the strategy. We might need to name this better

File: src/js/rtc_session.js [hunk 6]
```diff
@@ -361,32 +363,7 @@ export class AcceptState extends RTCSessionState {
...
-        }).then(() => {
-            rtcSession._sessionReport.setRemoteDescriptionFailure = false;
-            self._remoteDescriptionSet = true;
-            self._checkAndTransit();
-        }).catch(() => {
-            rtcSession._stopSession();
-            rtcSession._sessionReport.setRemoteDescriptionFailure = true;
-            self.transit(new FailedState(rtcSession, RTC_ERRORS.SET_REMOTE_DESCRIPTION_FAILURE));
-        });
+        self._rtcSession._strategy.acceptStateOnEnter(self, rtcSession);
```
Comment (Vignendra) [inline src/js/rtc_session.js:366]: same comment on the naming here.

File: src/js/rtc_session.js [hunk 7]
```diff
@@ -424,21 +401,23 @@ export class TalkingState extends RTCSessionState {
         this.transit(new DisconnectedState(this._rtcSession));
     }
     onIceStateChange(evt) {
-        this.logger.info('ICE Connection State: ', evt.currentTarget.iceConnectionState);
+        var iceState = this._rtcSession._strategy.onIceStateChange(evt, this._rtcSession._pc);
+        this.logger.info('ICE Connection State: ', iceState);
 
-        if (evt.currentTarget.iceConnectionState == ICE_CONNECTION_STATE.DISCONNECTED) {
+        if (iceState == ICE_CONNECTION_STATE.DISCONNECTED) {
```
Comment (Vignendra) [inline src/js/rtc_session.js:407]: Have tested these paths for citrix strategy

File: src/js/strategies/CCPInitiationStrategyInterface.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+
+export default class CCPInitiationStrategyInterface {
+    constructor() {
+        console.log("CCPInitiationStrategyInterface initialized");
+    }
+
+    getName(){}
```
Comment (Vignendra) [inline src/js/strategies/CCPInitiationStrategyInterface.js:7]: Any way we can ensure that all these method need to be overriden? Do we throw an expection in each of these methods saying functions needs to be overriden?

File: src/js/strategies/CitrixVDIStrategy.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    constructor() {
+        super();
+        if (!isCitrixWebRTCSupported()) {
+            throw new Error('Citrix WebRTC redirection feature is NOT supported!');
+        }
+        window.getCitrixWebrtcRedir = function() {
+            const registryValue = Promise.resolve(1);
```
Comment (Vignendra) [inline src/js/strategies/CitrixVDIStrategy.js:14]: whats with Promise.resolve(1)?

File: src/js/strategies/CitrixVDIStrategy.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                    reject(); })
+            })
+        }
+        window.CitrixWebRTC.initLog(global.connect.getLog());
+        console.log("CitrixVDIStrategy initialized");
+    }
+
+    // the following functions are rtc_peer_connection_factory related functions
+    // check if the browser supports early media connection
+    _isBrowserSupported(){
```
Comment (Vignendra) [inline src/js/strategies/CitrixVDIStrategy.js:28]: can we rename this to _isEarlyMediaConnectionSupported()?

File: src/js/strategies/CitrixVDIStrategy.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        // Citrix WebRTC SDK doesn't support early media connection
+        return false;
+    }
+
+    _createRtcPeerConnection(rtcPeerConnectionConfig, rtcPeerConnectionOptionalConfig) {
```
Comment (Vignendra) [inline src/js/strategies/CitrixVDIStrategy.js:33]: Thoughts on having rtcPeerConnectionConfigs defined within each strategy. This way we needn't keep configs common across strategy, if in the future we might need some config only for some specific strategy.

File: src/js/strategies/CitrixVDIStrategy.js [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        }, (error) => {
+            console.error(`ConnectRTC - rtc_session - AcceptState - setRemoteDescription - error:`, error);
+            rtcSession._stopSession();
+            rtcSession._sessionReport.setRemoteDescriptionFailure = true;
+            self.transit(new FailedState(rtcSession, RTC_ERRORS.SET_REMOTE_DESCRIPTION_FAILURE));
+        });
+    }
+
+    onIceStateChange(evt, _pc) {
+        return _pc.iceConnectionState;
```
Comment (Vignendra) [inline src/js/strategies/CitrixVDIStrategy.js:78]: Do we want to log this? or is the method you are calling logging it?

File: src/js/strategies/CitrixVDIStrategy.js [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+                mediaConstraints.audio = audioConstraints;
+            } else {
+                mediaConstraints.audio = true;
+            }
+        } else {
+            mediaConstraints.audio = false;
+        }
+    }
+
+    getName() {
```
Comment (Vignendra) [inline src/js/strategies/CitrixVDIStrategy.js:128]: getStrategyName?

File: src/js/strategies/StandardStrategy.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+            self._remoteDescriptionSet = true;
+            self._checkAndTransit();
+        }).catch(() => {
+            rtcSession._stopSession();
+            rtcSession._sessionReport.setRemoteDescriptionFailure = true;
+            self.transit(new FailedState(rtcSession, RTC_ERRORS.SET_REMOTE_DESCRIPTION_FAILURE));
+        });
+    }
+
+    onIceStateChange(evt, _pc) { // eslint-disable-line no-unused-vars
```
Comment (Vignendra) [inline src/js/strategies/StandardStrategy.js:60]: nit: can we remove this comment?

File: src/js/utils.js [hunk 15]
```diff
@@ -252,4 +252,16 @@ export function getChromeBrowserVersion(){
     } else {
         return -1;
     }
-}
\ No newline at end of file
+}
+
+export function isCitrixWebRTCSupported() {
```
Comment (Vignendra) [inline src/js/utils.js:257]: why is this part of common utils class? Can this be made part of CitrixStrategy?

File: src/js/rtc_const.js (hunk 3 above)
Comment (siwei-w) [reply src/js/rtc_const.js:37]: It won't affect non-citrix calls.
The non-citrix calls have the same value for these fields by default.
bundlePolicy is `balanced` by default. https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection#bundlepolicy
sdpSemantics is `unified` by default https://webrtc.org/getting-started/unified-plan-transition-guide#phase_3_switch_the_default
enableDtlsSrtp is `true` by default

File: src/js/strategies/CCPInitiationStrategyInterface.js (hunk 8 above)
Comment (siwei-w) [reply src/js/strategies/CCPInitiationStrategyInterface.js:7]: No, we don't throw any expections

File: src/js/strategies/CitrixVDIStrategy.js (hunk 9 above)
Comment (siwei-w) [reply src/js/strategies/CitrixVDIStrategy.js:14]: Citrix SDK needs to retrieve redirection status registry value. We need to define a method on window object that returns this status registry value, and  the return value needs to be a Promise of value 0, 1, or empty string.
We define Promise.resolve(1) here to indicate the redirection status is success.

File: src/js/rtc_peer_connection_factory.js (hunk 4 above)
Comment (siwei-w) [reply src/js/rtc_peer_connection_factory.js:10]: Yeah, we have reviewed the naming internally.

File: src/js/strategies/StandardStrategy.js (hunk 14 above)
Comment (siwei-w) [reply src/js/strategies/StandardStrategy.js:60]: No, build will fail

File: src/js/rtc_session.js (hunk 5 above)
Comment (siwei-w) [reply src/js/rtc_session.js]: Renamed

File: src/js/rtc_session.js (hunk 6 above)
Comment (siwei-w) [reply src/js/rtc_session.js]: Renamed

File: src/js/rtc_session.js (hunk 7 above)
Comment (siwei-w) [reply src/js/rtc_session.js:407]: Yes, tested

File: src/js/strategies/CitrixVDIStrategy.js (hunk 10 above)
Comment (siwei-w) [reply src/js/strategies/CitrixVDIStrategy.js]: Renamed

File: src/js/strategies/CitrixVDIStrategy.js [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    acceptStateOnEnter(self, rtcSession) {
+        console.log(`ConnectRTC - rtc_session - AcceptState - rtcSession._pc.setRemoteDescription`);
+
+        const answerSessionDescription = self._createSessionDescription({ type: 'answer', sdp: self._sdp });
+        console.log(`ConnectRTC - rtc_session - AcceptState - answerSessionDescription: `, answerSessionDescription);
+
+        rtcSession._pc.setRemoteDescription(answerSessionDescription, () => {
+            console.log(`ConnectRTC - rtc_session - AcceptState - rtcSession._pc.setRemoteDescription - success`);
```
[the remark this answers is not in the document]
Comment (siwei-w) [reply src/js/strategies/CitrixVDIStrategy.js]: Removed

File: src/js/strategies/CitrixVDIStrategy.js (hunk 13 above)
Comment (siwei-w) [reply src/js/strategies/CitrixVDIStrategy.js]: Renamed

File: src/js/utils.js (hunk 15 above)
Comment (siwei-w) [reply src/js/utils.js]: Moved

File: src/js/strategies/CitrixVDIStrategy.js (hunk 1 above)
Comment (siwei-w) [reply src/js/strategies/CitrixVDIStrategy.js]: I've removed the console logs here

File: src/js/strategies/CCPInitiationStrategyInterface.js (hunk 8 above)
Comment (siwei-w) [reply src/js/strategies/CCPInitiationStrategyInterface.js]: We will log in the console, we don't throw exceptions

File: src/js/strategies/CitrixVDIStrategy.js (hunk 11 above)
Comment (siwei-w) [reply src/js/strategies/CitrixVDIStrategy.js:40]: Didn't address this in this PR, we will pick up this later

File: src/js/strategies/CitrixVDIStrategy.js (hunk 12 above)
Comment (siwei-w) [reply src/js/strategies/CitrixVDIStrategy.js:78]: Checked with @yiming-amzn, this is not required.

File: package.json [hunk 17]
```diff
@@ -25,6 +25,7 @@
     "prepublish": "grunt build",
     "test": "grunt && nyc --reporter=html mocha --compilers js:babel-core/register test/unit"
   },
+  "_ucsdk_upgrade_note" : "Please follow the instruction here(https://tiny.amazon.com/18mszf5px/quipaTMFStep) to upgrade @citrix/ucsdk",
```
[the remark this answers is not in the document]
Comment (siwei-w) [reply package.json:28]: This is only to be used by Amazon internal developers for upgrading Citrix if necessary.

**Review by Vignendra — APPROVED** [review]

**Review by Vignendra — APPROVED** [review]

**Review by Vignendra — APPROVED** [review]

File: src/js/rtc_peer_connection_factory.js [hunk 18]
```diff
@@ -1,34 +1,41 @@
...
-        this._browserSupported = this._isBrowserSupported();
+        this._earlyMediaConnectionSupported = this._isEarlyMediaConnectionSupported();
         this._initializeWebSocketEventListeners();
         this._requestPeerConnection();
         this._networkConnectivityChecker();
+
+        this._logger.log("RTC.js is using " + strategy.getStrategyName());
     }
 
-    _isBrowserSupported() {
```
Comment (Akash1684) [inline src/js/rtc_peer_connection_factory.js:30]: Why we are replacing generic method of `browserSupport` with EaryMedia check?

**Review by Akash1684 — APPROVED** [review]
One minor query

**Pull request merged**